plenoptic/simulate/models/frontend.py

```diff
@@ -1,96 +1,513 @@
+    times that of the center Gaussian.
+
+    Parameters
+    ----------
+    kernel_size:
+        Shape of convolutional kernel.
+    on_center:
+        Dictates whether center is on or off; surround will be the opposite of center
+        (i.e. on-off or off-on).
+    width_ratio_limit:
+        Ratio of surround stdev over center stdev. Surround stdev will be clamped to
+        ratio_limit times center_std.
+    amplitude_ratio:
+        Ratio of center/surround amplitude. Applied before filter normalization.
+    center_std:
+        Standard deviation of circular Gaussian for center.
+    surround_std:
+        Standard deviation of circular Gaussian for surround. Must be at least
+        ratio_limit times center_std.
+    pad_mode:
+        Padding for convolution, defaults to "circular".
     """
 
-    def __init__(self, disk_mask=False, pretrained=False, requires_grad=True):
+    def __init__(
+        self,
+        kernel_size: Union[int, Tuple[int, int]],
+        on_center: bool = True,
+        width_ratio_limit: float = 4.0,
+        amplitude_ratio: float = 1.25,
+        center_std: float = 1.0,
+        surround_std: float = 4.0,
+        pad_mode: str = "circular",
+    ):
+        super().__init__()
+
+        assert width_ratio_limit > 1.0, "stdev of surround must be greater than center"
+        assert amplitude_ratio >= 1.0, "ratio of amplitudes must at least be 1."
+
+        self.on_center = on_center
+        self.kernel_size = kernel_size
+        self.width_ratio_limit = width_ratio_limit
+        self.register_buffer("amplitude_ratio", torch.tensor(amplitude_ratio))
+
+        self.center_std = nn.Parameter(torch.tensor(center_std))
+        self.surround_std = nn.Parameter(torch.tensor(surround_std))
+
+        self.pad_mode = pad_mode
+
+    @property
+    def filt(self) -> Tensor:
+        """Creates an on center/off surround, or off center/on surround conv filter"""
+        filt_center = circular_gaussian2d(self.kernel_size, self.center_std)
+        filt_surround = circular_gaussian2d(self.kernel_size, self.surround_std)
+        on_amp = self.amplitude_ratio
+
+        if self.on_center:  # on center, off surround
+            filt = on_amp * filt_center - filt_surround  # on center, off surround
+        else:  # off center, on surround
+            filt = on_amp * filt_surround - filt_center
+
+        filt = filt / filt.sum()
+
+        return filt
+
+    def _clamp_surround_std(self) -> Tensor:
+        """Clamps surround standard deviation to ratio_limit times center_std"""
+        return self.surround_std.clamp(
+            min=self.width_ratio_limit * float(self.center_std), max=None
+        )
+
+    def forward(self, x: Tensor) -> Tensor:
+        x = same_padding(x, self.kernel_size, pad_mode=self.pad_mode)
+        self._clamp_surround_std()  # clip the surround stdev
+        y = F.conv2d(x, self.filt, bias=None)
+        return y
```

Comment:
why doesn't `CenterSurround` have a `display_filters` method? Also, it feels like it should be possible to write one `display_filters` function that these all use (the only thing that really differs among them is the title).

Reply:
Your first sentence is resolved now that it was moved to `canonical_computations.filters.py`. It's true there's some redundancy there. The weights that need to be displayed are different w/ each model as well as their corresponding titles. I don't think it's so bad.